text/0000-enable-keyboard-avoiding.md

```diff
@@ -0,0 +1,63 @@
+- Start Date: 2013-07-23
+- RFC PR: (leave this empty)
+- React Navigation Issue: (leave this empty)
+
+# Summary
+
+Add enableKeyboardAvoiding to navigationConfig.
+
+# Basic example
+
+```
+createStackNavigator({...}, { enableKeyboardAvoiding: true });
+```
+
+# Motivation
+
+keyboardVerticalOffset is required if the KeyboardAvoidingView does not start from the top of the screen.
+For createStackNavigator or createMaterialTopTabNavigator there will headers and tabs.
+If developers need a KeyboardAvoidingView inside StackNavigator/TopTabNavigator, they cannot provide the offset.
+
+I've seen most of the projects use magic numbers (like 64 for ios, 54+StatusBar.currentHeight for Android, and special case for iPhone X, iPad, horizontal iPhone X...) to set the keyboardVerticalOffset. Which is not safe.
+
+If you are thinking why not put the KeyboardAvoidingView outside the StackNavigator. The answer sometimes we cannot, for example below structure:
+```
+<BottomTabNavigator>
+  <StackNavigator />
+</BottomTabNavigator>
+```
+We don't want bottom tabs to be "KeyboardAvoiding", we only want the StackNavigator to be "KeyboardAvoiding"
+```
+<BottomTabNavigator>
+  <KeyboardAvoidingView>
```

Comment:
it's actually possible to do this using the approach described in this section of the docs - you just need to expose the router and thread through the navigation prop on a wrapper component: https://reactnavigation.org/docs/en/common-mistakes.html#explicitly-rendering-more-than-one-navigator

by doing this, you could make a HOC to handle this for you! a dummy implementation that i haven't tested but should work fine:

```
function withKeyboardAvoidingView(Navigator) {
  class KeyboardAvoidingNavigator extends React.Component {
    static router = Navigator.router;

    render() {
      return <KeyboardAvoidingView><Navigator {...this.props} /></KeyboardAvoidingView>
    }
  }

  return KeyboardAvoidingNavigator;
}

// then later on
const StackNavigator = withKeyboardAvoidingView(createStackNavigator({ /* */ }));
```

what do you think? perhaps we need to document this pattern better

Reply:
@brentvatne sad to say this, but your example is not working: wrapping by navigator to that produces just a white screen. Added `enabled style={{flex: 1}}` renders content, but keyboard is not avoided.

It would be very great to handle such common case with the library.

Reply:
as i mentioned this was just a dummy implementation and i didn't test it, if you add flex: 1 and that works then great! it was actually close then :P as for it not having keyboard avoiding behavior, i can't say because clearly the keyboardavoidingview would be rendered so you'd have to see how that works

Reply:
Hi @brentvatne, so your solution is actually very close to my solution in real project.
https://github.com/yangyuan/react-native-social-networks/blob/master/src/app/components/navigation.tsx

1. I do consider this keyboardavoiding feature as a common requirements and should be take care by navigation frameworks.
2.  Above code works in currently version, will they work in future versions?
3. `style={{ flex: 1 }} behavior="padding"` seems to be required (I didn't do full investigation). But to understand the reason, the developers will have to know some details about the implementation of react-navigation. And not everyone can figure that out easily.

Reply:
my concern is that we probably don't want to require people to use this particular component - i don't think KeyboardAvoidingView is very well maintained or particularly good. until we have a very solid abstraction for this, why don't we document this approach and publish `createStackNavigatorKeyboardAvoiding` as a standalone package on npm?

Reply:
@brentvatne 

So.... if KeyboardAvoidingView is not well maintained, then I do agree that document this or publish a side-package is a better solution.

But just curious, what makes you think KeyboardAvoidingView is not well maintained? Did ReactNative community had some discussions about it or are they planning for some replacements?

Reply:
https://github.com/facebook/react-native/issues?q=KeyboardAvoidingView+is%3Aopen - nobody will fix these. looks at the git history for KeyboardAvoidingView and you will see that the almost all of the changes to it over the last couple years are codemods to update syntax, flow types, license, etc. https://medium.freecodecamp.org/how-to-make-your-react-native-app-respond-gracefully-when-the-keyboard-pops-up-7442c1535580 shows another approach. maybe people build their thing for this too.